sql/analyzer/resolve_variables.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2020 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analyzer
+
+import (
+	"strings"
+
+	"github.com/liquidata-inc/vitess/go/vt/sqlparser"
+
+	"github.com/liquidata-inc/go-mysql-server/sql"
+	"github.com/liquidata-inc/go-mysql-server/sql/expression"
+	"github.com/liquidata-inc/go-mysql-server/sql/parse"
+	"github.com/liquidata-inc/go-mysql-server/sql/plan"
+)
+
+// resolveSetVariables replaces SET @@var and SET @var expressions with appropriately resolved expressions for the
+// left-hand side, and evaluate the right-hand side where possible, including filling in defaults. Also validates that
+// system variables are known to the system.
+func resolveSetVariables(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, error) {
+	_, ok := n.(*plan.Set)
+	if !ok || n.Resolved() {
+		return n, nil
+	}
+
+	return plan.TransformExpressionsUp(n, func(e sql.Expression) (sql.Expression, error) {
+		sf, ok := e.(*expression.SetField)
+		if !ok {
+			return e, nil
+		}
+
+		varName := trimVarName(sf.Left.String())
+		setVal, err := getSetVal(ctx, varName, sf.Right)
+		if err != nil {
+			return nil, err
+		}
+
+		// For the left side of the SetField expression, we will attempt to resolve the variable being set. The rules to
+		// determine whether to treat a left-hand expression as a system var or something else are subtle. These are all
+		// the valid ways to assign to a system variable with session scope:
+		// SET SESSION sql_mode = 'TRADITIONAL';
+		// SET LOCAL sql_mode = 'TRADITIONAL';
+		// SET @@SESSION.sql_mode = 'TRADITIONAL';
+		// SET @@LOCAL.sql_mode = 'TRADITIONAL';
+		// SET @@sql_mode = 'TRADITIONAL';
+		// SET sql_mode = 'TRADITIONAL';
+		// These are all equivalent, and all distinct from setting a user variable with the same name:
+		// set @sql_mode = "abc"
+		if uc, ok := sf.Left.(*expression.UnresolvedColumn); ok {
+			if isSystemVariable(uc) {
+				// TODO: clean up distinction between system and user vars in this interface
+				typ, _ := ctx.Session.Get(varName)
+				if typ == sql.Null {
+					return nil, sql.ErrUnknownSystemVariable.New(varName)
+				}
+
+				// Special case: for system variables, MySQL allows naked strings (without quotes), which get interpreted as
```

Comment:
MySQL just loves to make rules to break later. If this is a thing in other databases then it's stupid there too lol